src/lib.rs

```diff
@@ -0,0 +1,379 @@
+    can_bump: bool,
+}
+
+struct AllocListIter<'a> {
+    alloc_list: &'a AllocList,
+    index: usize,
+}
+
+struct AllocListIterMut<'a> {
+    alloc_list: &'a mut AllocList,
+    index: usize,
+}
+
+impl<'a> Iterator for AllocListIterMut<'a> {
+    type Item = &'a mut Block;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // It's UB to call `.add` on a pointer past its allocation bounds, so we
+        // need to check that it's within range before turning it into a pointer
+        // and dereferencing it
+        let ulimit = self.alloc_list.start as usize + SIZE_ALLOC_TABLE;
+        let ptr = self.alloc_list.start as usize + (self.index * core::mem::size_of::<Block>());
+
+        self.index += 1;
+
+        if ptr >= ulimit {
+            return None;
+        }
+
+        unsafe { Some(&mut *(ptr as *mut Block)) }
+    }
+}
+
+impl<'a> Iterator for AllocListIter<'a> {
+    type Item = &'a Block;
+
+    fn next(&mut self) -> Option<&'a Block> {
+        // It's UB to call `.add` on a pointer past its allocation bounds, so we
+        // need to check that it's within range before turning it into a pointer
+        // and dereferencing it
+        let ulimit = self.alloc_list.start as usize + SIZE_ALLOC_TABLE;
+        let ptr = self.alloc_list.start as usize + (self.index * core::mem::size_of::<Block>());
+
+        self.index += 1;
+
+        if ptr >= ulimit {
+            return None;
+        }
+
+        let entry = unsafe { &*(ptr as *const Block) };
+        Some(&entry)
+    }
+}
+
+impl AllocList {
+    fn new() -> AllocList {
+        let raw = unsafe { libc::malloc(SIZE_ALLOC_TABLE as libc::size_t) } as *mut Block;
+        AllocList {
+            start: raw,
+            next_free: 0,
+            can_bump: true,
+        }
+    }
+
+    fn iter(&self) -> AllocListIter {
+        AllocListIter {
+            alloc_list: self,
+            index: 0,
+        }
+    }
+
+    fn iter_mut(&mut self) -> AllocListIterMut {
+        AllocListIterMut {
+            alloc_list: self,
+            index: 0,
+        }
+    }
+
+    /// Performs fast bump pointer insertion until the list is full, at which
+    /// point, insertion is O(n) while it linearly scans the list for the next
+    /// free entry.
+    fn insert(&mut self, ptr: usize, size: usize) {
+        if self.can_bump {
+            unsafe {
+                let next_ptr = self.start.add(self.next_free) as *mut Block;
+                if (next_ptr as usize) < self.start as usize + SIZE_ALLOC_TABLE {
+                    *next_ptr = Block::Entry(core::num::NonZeroUsize::new_unchecked(ptr), size);
+                    self.next_free += 1;
+                    return;
+                } else {
+                    self.can_bump = false;
+                }
+            }
+        }
+
+        // Slow path, we need to linearly scan for the next free block in the
+        // heap.
+        for block in self.iter_mut() {
+            match block {
```

Comment:
If you have a `_ => ()` case, you probably want `if let` (and it reduces indentation by 1 level, which is nice!).

Reply:
Fixed